src/pip/_internal/download.py

```diff
@@ -141,38 +147,141 @@ def user_agent():
 
+        # Early exit if we have cached credentials for this netloc
+        if username is not None and password is not None:
+            return url, username, password
+
+        # Find a matching index url for this request
+        index_url = self._get_index_url(url)
+        if index_url:
+            # Split the credentials from the url.
+            index_url, _, index_url_user_password = split_auth_netloc_from_url(
+                index_url)
+            logger.debug("Request for index url %s", index_url)
+
         # Use the credentials embedded in the url if we have none stored
-        if username is None:
+        if username is None or password is None:
             username, password = url_user_password
+            if username or password:
+                logger.debug("Found credentials in url for %s", netloc)
+            elif index_url:
+                username, password = index_url_user_password
+                if username or password:
+                    logger.debug("Found credentials in index url for %s",
+                                 index_url)
 
         # Get creds from netrc if we still don't have them
-        if username is None and password is None:
-            netrc_auth = get_netrc_auth(req.url)
-            username, password = netrc_auth if netrc_auth else (None, None)
+        if allow_netrc and username is None and password is None:
+            netrc_url, netrc_auth = None, None
+            if index_url:
+                netrc_url = index_url
+                netrc_auth = get_netrc_auth(index_url)
+            if not netrc_auth:
+                netrc_url = original_url
+                netrc_auth = get_netrc_auth(original_url)
+            if netrc_auth:
+                logger.debug("Found credentials in netrc for %s", netrc_url)
+                username, password = netrc_auth
+
+        # If we don't have a password and keyring is available, use it.
+        if allow_keyring and password is None and keyring is not None:
+            try:
+                kr_auth = None
+                if index_url:
+                    kr_auth = _get_keyring_auth(index_url, username)
+                if not kr_auth:
+                    kr_auth = _get_keyring_auth(netloc, username)
+                if kr_auth:
+                    username, password = kr_auth
+            except Exception as exc:
+                logger.warning("Keyring is skipped due to an exception: %s",
+                               str(exc))
 
         if username or password:
             # Store the username and password
             self.passwords[netloc] = (username, password)
 
+        return url, username, password
+
+    def __call__(self, req):
+        # Get credentials for this request
+        url, username, password = self._get_url_and_credentials(req.url)
+
+        # Set the url of the request to the url without any credentials
+        req.url = url
+
+        if username or password:
```

Comment:
I'll just note that here also only one of `username` and `password` needs to be truthy, whereas above you required both to be non-`None` in order to use what was in the cache. Again, this would be good to clarify.